rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java

```diff
@@ -0,0 +1,256 @@
+
+                Cursor maybeReplacementRoot = null, maybeReplacementScope = null;
+                for (Iterator<Cursor> it = getCursor().getPathAsCursors(); it.hasNext(); ) {
+                    Cursor next = it.next();
+                    Object value = next.getValue();
+                    if (value instanceof J.Binary) {
+                        J.Binary binary = (J.Binary) value;
+                        if (binary.getOperator() == J.Binary.Type.And) {
+                            maybeReplacementScope = next;
+                        } else {
+                            break;
+                        }
+                    } else if (value instanceof Statement) {
+                        maybeReplacementRoot = next;
+                        break;
+                    }
+                }
+
+                if (maybeReplacementScope != null && maybeReplacementRoot != null) {
+                    J root = maybeReplacementRoot.getValue();
+                    Set<J> contexts = new HashSet<>();
+                    if (root instanceof J.If) {
+                        contexts.add(((J.If) root).getThenPart());
+                    }
+                    contexts.add(maybeReplacementScope.getValue());
+
+                    InstanceOfPatternReplacements replacements = maybeReplacementRoot
+                            .computeMessageIfAbsent("flowTypeScope", k -> new InstanceOfPatternReplacements(root));
+                    replacements.register(instanceOf, contexts);
+
+                    if (replacementVisitor == null) {
+                        replacementVisitor = new UseInstanceOfPatternMatching();
+                        doAfterVisit(replacementVisitor);
+                    }
+                    replacementVisitor.register(replacements);
+                }
+                return instanceOf;
+            }
+
+            @Override
+            public J.TypeCast visitTypeCast(J.TypeCast typeCast, ExecutionContext executionContext) {
+                typeCast = super.visitTypeCast(typeCast, executionContext);
+                InstanceOfPatternReplacements replacements = getCursor().getNearestMessage("flowTypeScope");
+                if (replacements != null) {
+                    replacements.register(typeCast, getCursor());
+                }
+                return typeCast;
+            }
+        };
+    }
+
+    @Data
+    private static class InstanceOfPatternReplacements {
+        private final J root;
+        private final Map<Expression, J.InstanceOf> instanceOfs = new HashMap<>();
+        private final Map<J.InstanceOf, Set<J>> contexts = new HashMap<>();
+        private final Map<J.TypeCast, J.InstanceOf> replacements = new HashMap<>();
+
+        public void register(J.InstanceOf instanceOf, Set<J> contexts) {
+            Expression expression = instanceOf.getExpression();
+            Optional<Expression> existing = instanceOfs.keySet().stream().filter(e -> SemanticallyEqual.areEqual(e, expression)).findAny();
```

Comment:
Whether all semantically equal expressions or only simple variable references should be refactored by this recipe or not should probably be controlled by an option, as more complex expressions could have side effects. But I think that would be a very weird coding style...

Reply:
Without saying that this usage is problematic or anything, note that `SemanticallyEqual` is not really comprehensive or even entirely well defined. There are lots of things that could be considered equal which are not implemented.